### PR TITLE
secp256k1: Test consistency cleanup and rework.

### PR DESCRIPTION
**Pull request by davecgh**
~**This requires #2886**~.

This cleans up a bunch of tests and benchmarks to make them more consistent with modern practices in the code and also moves the ones that only apply when accessing the package via the standard library elliptic curve adaptor code to files dedicated to that purpose.  It also adds several new tests for various edge cases and ensures the randomized testing uses new seeds on each invocation to ensure new random values are tested each run.

It is split up into a series of commits to help ease the review process.

The following is a high level overview of the changes:

- Corrects several comments
- Makes several benchmarks more consistent in terms of  resetting the timer after the initial setup and reporting memory allocations
- Consolidates test logic that ensures test point in Jacobian coordinates are valid group elements
- Adds additional assurances that test points in affine coordinates are valid group elements
- Cleans up affine addition tests and moves them `ellipticadaptor_test.go`
- Cleans up affine double tests and moves them `ellipticadaptor_test.go`
- Cleans up key generation tests and moves them `privkey_test.go`
  - Removes the curve-based indirection since it only works with secp256k1
- Cleans up affine scalar point multiplication tests and moves them `ellipticadaptor_test.go`
  - Adds some additional edge cases
- Cleans up affine scalar base point multiplication tests
  - Adds some additional edge cases
 - Cleans up randomized scalar base point multiplication tests and moves them `ellipticadaptor_test.go`
   - Adds a `randBytes` convenience func to acquire a specified number of random bytes from a given rng
   - Uses a new seed for each invocation to ensure new random values are tested each run 
 - Cleans up private key serialization tests
 - Cleans up Jacobian addition tests
 - Cleans up Jacobian double tests
 - Adds a test helper for generating random mod n scalars from a given rng
 - Adds tests for Jacobian scalar base point multiplication
 - Reworks randomized Jacobian scalar multiplication tests
   - The points are no longer converted to affine with each iteration which ensures z values other than 1 are tested
   - Each iteration now ensures calculating the negative version of the   point and adding it results in the point at infinity
   - The check to ensure the same final point was calculated is now done outside of the loop to speed up the test
- Other miscellaneous cleanup that is described in the individual commit messages

### CODE REVIEW

File: dcrec/secp256k1/curve_test.go [hunk 1]
```diff
@@ -746,6 +710,93 @@ func TestSplitKRand(t *testing.T) {
...
+	// First, calculate a "chained" point by starting with the base (generator)
+	// point and then consecutively multiply the resulting points by a series of
+	// random scalars.
+	//
+	// Then, multiply the base point by the product of all of the random scalars
+	// and ensure the "chained" point matches.
+	//
+	// In other words:
+	//
+	// k[n]*(...*(k[2]*(k[1]*(k[0]]*G)))) = (k[0]*k[1]*k[2]*...*k[n])*G
```
Comment (davecgh) [inline dcrec/secp256k1/curve_test.go:752]:
```suggestion
	// k[n]*(...*(k[2]*(k[1]*(k[0]*G)))) = (k[0]*k[1]*k[2]*...*k[n])*G
```

File: dcrec/secp256k1/curve_test.go [hunk 2]
```diff
@@ -854,21 +710,91 @@ func TestSplitKRand(t *testing.T) {
...
 	}
-	pub := priv.PubKey()
-	if !isOnCurve(&pub.x, &pub.y) {
-		t.Errorf("%s: public key invalid: %s", tag, err)
+
+	// The overall idea is to compute the same point different ways.  The
+	// strategy uses two properties:
+	//
+	// 1) Compatibility of scalar multiplication with field multiplication
+	// 2) A point added to its negation is the point at infinitity (P+(-P) = ∞)
```
Comment (rstaudt2) [inline dcrec/secp256k1/curve_test.go:741]:
```suggestion
	// 2) A point added to its negation is the point at infinity (P+(-P) = ∞)
```

File: dcrec/secp256k1/field.go [hunk 3]
```diff
@@ -1156,7 +1156,7 @@ func (f *FieldVal) SquareRootVal(val *FieldVal) bool {
 	// 10111111 11111111 11111111 00001100
 	//
 	// Notice that can be broken up into three windows of consecutive 1s (in
-	// order of least to most signifcant) as:
+	// order of least to most significant) as:
```
Comment (rstaudt2) [inline dcrec/secp256k1/field.go:1159]: Copyright year could be updated in this file as well.

**Review by rstaudt2 — APPROVED** [review]
Looks good, updates make the tests more consistent and easier to follow!

File: dcrec/secp256k1/field.go (hunk 3 above)
Comment (davecgh) [reply dcrec/secp256k1/field.go:1159]: Good call.  I'm pretty sure I did it in a later commit, but it should be in this one.  Done.

File: dcrec/secp256k1/curve_test.go (hunk 2 above)
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go:741]: Fixed.

**Review by dajohi — APPROVED** [review]

File: dcrec/secp256k1/ellipticadaptor_test.go [hunk 4]
```diff
@@ -1,65 +1,421 @@
...
+			continue
+		}
+		if !isValidAffinePoint(x3, y3) {
+			t.Errorf("%s: expected point is not on curve", test.name)
+			continue
+		}
+
+		// Add the two points and ensure the result matches expected.
+		rx, ry := curve.Add(x1, y1, x2, y2)
+		if rx.Cmp(x3) != 00 || ry.Cmp(y3) != 0 {
```
Comment (JoeGruffins) [inline dcrec/secp256k1/ellipticadaptor_test.go:125]: `00` could just be `0`? Unsure if this has a special meaning.

**Review by JoeGruffins — APPROVED** [review]
```
$ go test -bench=.
goos: linux
goarch: amd64
pkg: github.com/decred/dcrd/dcrec/secp256k1/v4
cpu: AMD Ryzen 9 3900XT 12-Core Processor           
BenchmarkAddNonConst-24                          3585380               306.6 ns/op             0 B/op          0 allocs/op
BenchmarkAddNonConstNotZOne-24                   1452766               796.9 ns/op             0 B/op          0 allocs/op
BenchmarkScalarBaseMultNonConst-24                 58899             17878 ns/op               0 B/op          0 allocs/op
BenchmarkScalarMultNonConst-24                     12162             98178 ns/op             720 B/op         11 allocs/op
BenchmarkNAF-24                                 17839897                63.77 ns/op            0 B/op          0 allocs/op
BenchmarkPubKeyDecompress-24                      127308              9277 ns/op               0 B/op          0 allocs/op
BenchmarkParsePubKeyCompressed-24                 126291              9406 ns/op              80 B/op          1 allocs/op
BenchmarkParsePubKeyUncompressed-24              3375386               361.2 ns/op            80 B/op          1 allocs/op
BenchmarkScalarBaseMultAdaptor-24                  40498             28251 ns/op             224 B/op          5 allocs/op
BenchmarkScalarBaseMultLargeAdaptor-24             37504             30458 ns/op             400 B/op          8 allocs/op
BenchmarkScalarMultAdaptor-24                      10488            110865 ns/op            1009 B/op         18 allocs/op
BenchmarkFieldNormalize-24                      60825061                16.67 ns/op            0 B/op          0 allocs/op
BenchmarkFieldSqrt-24                             130118              8961 ns/op               0 B/op          0 allocs/op
BenchmarkBigSqrt-24                                49902             22461 ns/op            1017 B/op         91 allocs/op
BenchmarkFieldIsGtOrEqPrimeMinusOrder-24        189939106                6.072 ns/op           0 B/op          0 allocs/op
BenchmarkBigIsGtOrEqPrimeMinusOrder-24           9318628               146.0 ns/op            96 B/op          2 allocs/op
BenchmarkBigIntModN-24                           3418603               376.6 ns/op            72 B/op          2 allocs/op
BenchmarkModNScalar-24                          107776508               10.58 ns/op            0 B/op          0 allocs/op
BenchmarkBigIntZero-24                          485790676                2.102 ns/op           0 B/op          0 allocs/op
BenchmarkModNScalarZero-24                      1000000000               0.4864 ns/op          0 B/op          0 allocs/op
BenchmarkBigIntIsZero-24                        1000000000               0.2550 ns/op          0 B/op          0 allocs/op
BenchmarkModNScalarIsZero-24                    1000000000               0.4916 ns/op          0 B/op          0 allocs/op
BenchmarkBigIntEquals-24                        214368002                5.438 ns/op           0 B/op          0 allocs/op
BenchmarkModNScalarEquals-24                    450851756                2.337 ns/op           0 B/op          0 allocs/op
BenchmarkBigIntAddModN-24                        2934122               428.4 ns/op           128 B/op          2 allocs/op
BenchmarkModNScalarAdd-24                       89115290                13.18 ns/op            0 B/op          0 allocs/op
BenchmarkBigIntMulModN-24                        1959603               619.4 ns/op           176 B/op          2 allocs/op
BenchmarkModNScalarMul-24                        5727213               189.8 ns/op             0 B/op          0 allocs/op
BenchmarkBigIntSquareModN-24                     1926706               618.9 ns/op           176 B/op          2 allocs/op
BenchmarkModNScalarSquare-24                     5744864               193.1 ns/op             0 B/op          0 allocs/op
BenchmarkBigIntNegateModN-24                     3064765               390.0 ns/op            72 B/op          2 allocs/op
BenchmarkModNScalarNegate-24                    197936019                5.791 ns/op           0 B/op          0 allocs/op
BenchmarkBigIntInverseModN-24                    1000000              1260 ns/op             576 B/op         12 allocs/op
BenchmarkModNScalarInverse-24                    1000000              1008 ns/op             472 B/op         12 allocs/op
BenchmarkBigIntIsOverHalfOrder-24               254648517                4.333 ns/op           0 B/op          0 allocs/op
BenchmarkModNScalarIsOverHalfOrder-24           242337598                4.852 ns/op           0 B/op          0 allocs/op
PASS
ok      github.com/decred/dcrd/dcrec/secp256k1/v4       54.926s
```

File: dcrec/secp256k1/ellipticadaptor_test.go (hunk 4 above)
Comment (davecgh) [reply dcrec/secp256k1/ellipticadaptor_test.go]: Indeed.  Updated.
